File: server/src/routes/profiles.js
const express = require('express');
const db = require('../database');
const { authMiddleware } = require('../middleware/auth');
const { subscriptionMiddleware } = require('../middleware/subscription');

const router = express.Router();

let websocketServer = null;

function setWebSocketServer(wsServer) {
  websocketServer = wsServer;
}

function broadcastToUser(userId, event, data) {
  if (!websocketServer) return;
  
  const connections = db.getConnectionsByUserId(userId);
  connections.forEach(conn => {
    websocketServer.clients.forEach(client => {
      if (client.socketId === conn.socket_id && client.readyState === 1) {
        client.send(JSON.stringify({ event, data }));
      }
    });
  });
}

router.get('/', authMiddleware, (req, res) => {
  try {
    const { path } = req.query;
    
    if (path) {
      const profile = db.getProfileByUserIdAndPath(req.user.id, path);
      if (!profile) {
        return res.status(404).json({ error: 'File not found' });
      }
      return res.json({
        files: [{
          file_path: profile.file_path,
          content: profile.content,
          version: profile.version,
          updated_at: profile.updated_at
        }]
      });
    }
    
    const profiles = db.getProfilesByUserId(req.user.id);
    res.json({
      files: profiles.map(p => ({
        file_path: p.file_path,
        content: p.content,
        version: p.version,
        updated_at: p.updated_at
      }))
    });
  } catch (error) {
    console.error('Get profiles error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/', authMiddleware, subscriptionMiddleware, (req, res) => {
  try {
    const { file_path, content, version } = req.body;
    
    if (!file_path || content === undefined) {
      return res.status(400).json({ error: 'file_path and content are required' });
    }
    
    const result = db.createOrUpdateProfile(req.user.id, file_path, content, version || 0);
    
    if (result.conflict) {
      return res.status(409).json({
        error: 'Version conflict',
        code: 'VERSION_CONFLICT',
        latest_content: result.existing.content,
        latest_version: result.existing.version
      });
    }
    
    broadcastToUser(req.user.id, 'file_updated', {
      file_path: result.file_path,
      version: result.version
    });
    
    res.json({
      file_path: result.file_path,
      version: result.version,
      updated_at: result.updated_at
    });
  } catch (error) {
    console.error('Create/update profile error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/sync', authMiddleware, subscriptionMiddleware, (req, res) => {
  try {
    const { since } = req.query;
    
    const profiles = db.getProfilesUpdatedAfter(req.user.id, since || '0');
    
    res.json({
      files: profiles.map(p => ({
        file_path: p.file_path,
        content: p.content,
        version: p.version,
        updated_at: p.updated_at
      })),
      server_time: new Date().toISOString()
    });
  } catch (error) {
    console.error('Sync profiles error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;
module.exports.setWebSocketServer = setWebSocketServer;

File: plugins/openclaw/src/profiles.py
import requests


class ProfilesClient:
    """Profiles API 客户端"""
    
    def __init__(self, cloud_url: str, token: str = None):
        self.cloud_url = cloud_url.rstrip('/')
        self.token = token
    
    def _get_headers(self) -> dict:
        """获取请求头"""
        headers = {'Content-Type': 'application/json'}
        if self.token:
            headers['Authorization'] = f'Bearer {self.token}'
        return headers
    
    def set_token(self, token: str):
        """设置 token"""
        self.token = token
    
    def get_profiles(self, path: str = None) -> dict:
        """获取 profiles
        
        Args:
            path: 可选的文件路径
            
        Returns:
            包含 files 列表的字典
        """
        url = f"{self.cloud_url}/api/profiles"
        if path:
            url += f"?path={path}"
        
        response = requests.get(url, headers=self._get_headers())
        
        if response.status_code == 200:
            return response.json()
        elif response.status_code == 404:
            return {'files': []}
        else:
            error = response.json().get('error', 'Unknown error')
            raise Exception(f"Get profiles failed: {error}")
    
    def upload_profile(self, file_path: str, content: str, version: int) -> dict:
        """上传 profile
        
        Args:
            file_path: 文件路径
            content: 文件内容
            version: 当前版本号
            
        Returns:
            成功时返回 {file_path, version, updated_at}
            冲突时抛出异常
        """
        url = f"{self.cloud_url}/api/profiles"
        data = {
            'file_path': file_path,
            'content': content,
            'version': version
        }
        
        response = requests.post(url, json=data, headers=self._get_headers())
        
        if response.status_code == 200:
            return response.json()
        elif response.status_code == 409:
            result = response.json()
            raise ConflictError(
                result.get('latest_content', ''),
                result.get('latest_version', 0)
            )
        elif response.status_code == 403:
            error = response.json().get('error', 'Subscription required')
            raise Exception(f"Upload failed: {error}")
        else:
            error = response.json().get('error', 'Unknown error')
            raise Exception(f"Upload failed: {error}")
    
    def sync_profiles(self, since: str = '0') -> dict:
        """增量同步 profiles
        
        Args:
            since: 时间戳
            
        Returns:
            包含 files 列表和 server_time 的字典
        """
        url = f"{self.cloud_url}/api/profiles/sync?since={since}"
        
        response = requests.get(url, headers=self._get_headers())
        
        if response.status_code == 200:
            return response.json()
        elif response.status_code == 403:
            error = response.json().get('error', 'Subscription required')
            raise Exception(f"Sync failed: {error}")
        else:
            error = response.json().get('error', 'Unknown error')
            raise Exception(f"Sync failed: {error}")


class ConflictError(Exception):
    """版本冲突异常"""
    
    def __init__(self, latest_content: str, latest_version: int):
        self.latest_content = latest_content
        self.latest_version = latest_version
        super().__init__(f"Version conflict: latest version is {latest_version}")
